cpp/src/arrow/extension/variable_shape_tensor.cc

```diff
@@ -0,0 +1,263 @@
+  }
+  auto value_type = storage_type->field(1)->type()->field(0)->type();
+  const size_t ndim =
+      std::static_pointer_cast<FixedSizeListType>(storage_type->field(0)->type())
+          ->list_size();
+
+  rj::Document document;
+  if (document.Parse(serialized_data.data(), serialized_data.length()).HasParseError()) {
+    return Status::Invalid("Invalid serialized JSON data: ", serialized_data);
+  }
+
+  std::vector<int64_t> permutation;
+  if (document.HasMember("permutation")) {
+    for (auto& x : document["permutation"].GetArray()) {
+      permutation.emplace_back(x.GetInt64());
+    }
+    if (permutation.size() != ndim) {
+      return Status::Invalid("Invalid permutation");
+    }
+  }
+  std::vector<std::string> dim_names;
+  if (document.HasMember("dim_names")) {
+    for (auto& x : document["dim_names"].GetArray()) {
+      dim_names.emplace_back(x.GetString());
+    }
+    if (dim_names.size() != ndim) {
+      return Status::Invalid("Invalid dim_names");
+    }
+  }
+
+  std::vector<std::optional<int64_t>> uniform_shape;
+  if (document.HasMember("uniform_shape")) {
+    for (auto& x : document["uniform_shape"].GetArray()) {
+      if (x.IsNull()) {
+        uniform_shape.emplace_back(std::nullopt);
+      } else {
+        uniform_shape.emplace_back(x.GetInt64());
+      }
+    }
+    if (uniform_shape.size() > ndim) {
+      return Status::Invalid("Invalid uniform_shape");
+    }
+  }
+
+  return variable_shape_tensor(value_type, static_cast<uint32_t>(ndim), permutation,
+                               dim_names, uniform_shape);
+}
+
+std::shared_ptr<Array> VariableShapeTensorType::MakeArray(
+    std::shared_ptr<ArrayData> data) const {
+  DCHECK_EQ(data->type->id(), Type::EXTENSION);
+  DCHECK_EQ("arrow.variable_shape_tensor",
+            static_cast<const ExtensionType&>(*data->type).extension_name());
+  return std::make_shared<ExtensionArray>(data);
+}
+
+Result<std::shared_ptr<Tensor>> VariableShapeTensorType::GetTensor(
+    const std::shared_ptr<ExtensionScalar>& scalar) const {
+  const auto tensor_scalar = internal::checked_pointer_cast<StructScalar>(scalar->value);
+  const auto fields = this->storage_type()->fields();
+
+  ARROW_ASSIGN_OR_RAISE(const auto shape_scalar, tensor_scalar->field(0));
+  ARROW_ASSIGN_OR_RAISE(const auto data, tensor_scalar->field(1));
+  const auto shape_array =
+      std::static_pointer_cast<FixedSizeListScalar>(shape_scalar)->value;
+
+  std::vector<int64_t> shape;
+  for (uint32_t j = 0; j < this->ndim(); ++j) {
+    ARROW_ASSIGN_OR_RAISE(auto size, shape_array->GetScalar(j));
+    shape.push_back(
+        static_cast<int64_t>(std::static_pointer_cast<UInt32Scalar>(size)->value));
+  }
+
+  // TODO: optimize ComputeStrides for non-uniform tensors
+  std::vector<int64_t> strides;
+  ARROW_CHECK_OK(
+      internal::ComputeStrides(this->value_type(), shape, this->permutation(), &strides));
+
+  const auto array = std::static_pointer_cast<FixedSizeListScalar>(data)->value;
```

Comment:
It's not a fixed-size list, is it? :-)
Also, can use a const ref.
```suggestion
  const auto& array = checked_cast<const BaseListScalar&>(*data)->value;
```